File: sdk/noComment/HyperDbgDev/hyperdbg/hprdbgctrl/code/debugger/transparency/gaussian-rng.cpp
#include "pch.h"
double Median(vector<double> Cases) {
  size_t Size = Cases.size();
  if (Size == 0) {
    return 0; 
  } else {
    sort(Cases.begin(), Cases.end());
    if (Size % 2 == 0) {
      return (Cases[Size / 2 - 1] + Cases[Size / 2]) / 2;
    } else {
      return Cases[Size / 2];
    }
  }
}

template <typename T> T Average(const vector<T> &vec) {
  size_t Sz;
  T Mean;
  Sz = vec.size();
  if (Sz == 1)
    return 0.0;
  Mean = std::accumulate(vec.begin(), vec.end(), 0.0) / Sz;
  return Mean;
}

template <typename T> T CalculateStandardDeviation(const std::vector<T> &v) {
  double Sum, Mean, SqSum, Stdev;
  Sum = std::accumulate(v.begin(), v.end(), 0.0);
  Mean = Sum / v.size();
  SqSum = std::inner_product(v.begin(), v.end(), v.begin(), 0.0);
  Stdev = std::sqrt(SqSum / v.size() - Mean * Mean);
  return Stdev;
}

double MedianAbsoluteDeviationTest(vector<double> Data) {
  double MedianData;
  double Mad;
  MedianData = Median(Data);
  for (int i = 0; i < Data.size(); i++) {
    Data[i] = abs(Data[i] - MedianData);
  }
  Mad = 1.4826 * Median(Data);
  return Mad;
}

double Randn(double mu, double sigma) {
  double U1, U2, W, mult;
  static double X1, X2;
  static int call = 0;
  if (call == 1) {
    call = !call;
    return (mu + sigma * (double)X2);
  }
  do {
    U1 = -1 + ((double)rand() / RAND_MAX) * 2;
    U2 = -1 + ((double)rand() / RAND_MAX) * 2;
    W = pow(U1, 2) + pow(U2, 2);
  } while (W >= 1 || W == 0);
  mult = sqrt((-2 * log(W)) / W);
  X1 = U1 * mult;
  X2 = U2 * mult;
  call = !call;
  return (mu + sigma * (double)X1);
}

VOID GuassianGenerateRandom(vector<double> Data, UINT64 *AverageOfData,
                            UINT64 *StandardDeviationOfData,
                            UINT64 *MedianOfData) {
  vector<double> FinalData;
  int CountOfOutliers = 0;
  double Medians;
  double Mad;
  double StandardDeviation;
  double DataAverage;
  double DataMedian;
  vector<double> OriginalData = Data;
  vector<double> ChangableData = std::move(Data);
  Mad = MedianAbsoluteDeviationTest(ChangableData);
  Medians = Median(OriginalData);
  for (auto item : OriginalData) {
    if (item > (3 * Mad) + Medians || item < -(3 * Mad) + Medians) {
      CountOfOutliers++;
    } else {
      FinalData.push_back(item);
    }
  }
  StandardDeviation = CalculateStandardDeviation(FinalData);
  DataAverage = Average(FinalData);
  DataMedian = Median(FinalData);
  *AverageOfData = (UINT64)DataAverage;
  *StandardDeviationOfData = (UINT64)StandardDeviation + 5;
  *MedianOfData = (UINT64)DataMedian;
}

VOID TestGaussianFromFile() {
  vector<double> MyVector;
  UINT64 AverageOfData;
  UINT64 StandardDeviationOfData;
  UINT64 MedianOfData;
  std::ifstream file("C:\\Users\\sina\\Desktop\\r.txt");
  if (file.is_open()) {
    std::string line;
    while (std::getline(file, line)) {
      MyVector.push_back(stod(line.c_str()));
    }
    file.close();
    GuassianGenerateRandom(MyVector, &AverageOfData, &StandardDeviationOfData,
                           &MedianOfData);
  }
}

File: debugger/transparency/gaussian_rng.go
package transparency

import (
	"math"
	"math/rand"
	"sort"
	"sync"
)

const TestCount = 1000

func Median(cases []float64) float64 {
	size := len(cases)
	if size == 0 {
		return 0
	}
	sorted := make([]float64, size)
	copy(sorted, cases)
	sort.Float64s(sorted)

	if size%2 == 0 {
		return (sorted[size/2-1] + sorted[size/2]) / 2
	}
	return sorted[size/2]
}

func Average(vec []float64) float64 {
	size := len(vec)
	if size == 0 || size == 1 {
		return 0
	}
	var sum float64
	for _, v := range vec {
		sum += v
	}
	return sum / float64(size)
}

func CalculateStandardDeviation(v []float64) float64 {
	if len(v) == 0 {
		return 0
	}
	var sum, sqSum float64
	for _, val := range v {
		sum += val
	}
	mean := sum / float64(len(v))
	for _, val := range v {
		sqSum += val * val
	}
	variance := sqSum/float64(len(v)) - mean*mean
	if variance < 0 {
		variance = 0
	}
	return math.Sqrt(variance)
}

func MedianAbsoluteDeviationTest(data []float64) float64 {
	medianData := Median(data)
	for i := range data {
		data[i] = math.Abs(data[i] - medianData)
	}
	return 1.4826 * Median(data)
}

var (
	rngX1   float64
	rngX2   float64
	rngCall int
	rngMu   sync.Mutex
)

func Randn(mu, sigma float64) float64 {
	rngMu.Lock()
	defer rngMu.Unlock()

	if rngCall == 1 {
		rngCall = 0
		return mu + sigma*rngX2
	}

	for {
		u1 := -1 + (rand.Float64() * 2)
		u2 := -1 + (rand.Float64() * 2)
		w := u1*u1 + u2*u2

		if w >= 1 || w == 0 {
			continue
		}

		mult := math.Sqrt((-2 * math.Log(w)) / w)
		rngX1 = u1 * mult
		rngX2 = u2 * mult

		rngCall = 1
		return mu + sigma*rngX1
	}
}

type GaussianResult struct {
	Average           uint64
	StandardDeviation uint64
	Median            uint64
}

func GuassianGenerateRandom(data []float64) GaussianResult {
	var finalData []float64
	countOfOutliers := 0

	changableData := make([]float64, len(data))
	copy(changableData, data)

	mad := MedianAbsoluteDeviationTest(changableData)
	medians := Median(data)

	for _, item := range data {
		if item > (3*mad)+medians || item < -(3*mad)+medians {
			countOfOutliers++
		} else {
			finalData = append(finalData, item)
		}
	}

	standardDeviation := CalculateStandardDeviation(finalData)
	dataAverage := Average(finalData)
	dataMedian := Median(finalData)

	return GaussianResult{
		Average:           uint64(dataAverage),
		StandardDeviation: uint64(standardDeviation) + 5,
		Median:            uint64(dataMedian),
	}
}
